### PR TITLE
Fix float16 support for causal language models

### PR DESCRIPTION
**Pull request by danieldk**
Fix several dtype issues:

- When applying the logits mask, ensure that it is the right type.
- Use the minimum value for the dtype when applying the logits mask to avoid NaNs (as a result of -Infs).
- Convert rotary embeddings to the model's dtype.

The change also adds some tests for the float16 support. The tests are restricted to causal language models for now, since we primarily want to check that the models don't fail due to incorrect data types.

Furthermore, we now mask the logits before applying the test, to avoid large variances compared to HF test vectors for masked positions.

While at it, `assert_tensor_eq` and `check_causal_lm` are converted to macros, so that we can also use relative tolerances in the future. This is currently not used, but I experimented with relative tolerances while preparing this change and it would be a bit of a shame to let it whither.

### CODE REVIEW

File: oxidized-transformers/src/util/mod.rs [hunk 1]
```diff
@@ -50,9 +52,24 @@ pub(crate) mod tests {
...
     }
 
-    pub(crate) fn assert_tensor_eq<T>(a: impl IntoArrayD<T>, b: impl IntoArrayD<T>, epsilon: T)
-    where
-        T: AbsDiffEq<Epsilon = T> + Clone + Debug,
+    /// Check that two tensors are equal with the given absolute (`epsilon`)
+    /// and relative (`max_relative`) tolerances.
+    macro_rules! assert_tensor_eq {
+        ($lhs:expr, $rhs:expr $(, $opt:ident = $val:expr)*) => {
+            crate::util::tests::assert_tensor_eq_($lhs, $rhs, Relative::default()$(.$opt($val))*)
```
Comment (shadeMe) [inline oxidized-transformers/src/util/mod.rs]:
```suggestion
            crate::util::tests::assert_tensor_eq_($lhs, $rhs, approx::Relative::default()$(.$opt($val))*)
```

File: oxidized-transformers/src/util/mod.rs [hunk 2]
```diff
@@ -50,9 +52,24 @@ pub(crate) mod tests {
...
+        };
+        ($lhs:expr, $rhs:expr $(, $opt:ident = $val:expr)*,) => {
+            crate::util::tests::assert_tensor_eq_($lhs, $rhs, approx::Relative::default()$(.$opt($val))*)
+        };
```
Comment (shadeMe) [inline oxidized-transformers/src/util/mod.rs:63]: Why do we need a second definition with a trailing comma?

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: oxidized-transformers/src/util/mod.rs (hunk 2 above)
Comment (danieldk) [reply oxidized-transformers/src/util/mod.rs:63]: In case you write

```
assert_tensor_eq!(
    lhs,
    rhs,
    epsilon = 1e-4,
)
```

(Note the trailing comma after epsilon.)

File: oxidized-transformers/src/util/mod.rs (hunk 1 above)
Comment (danieldk) [reply oxidized-transformers/src/util/mod.rs]: Fixed.

**Review by shadeMe — APPROVED** [review]
